README.md
# casa.squid.jack

<!-- badges -->
[![cljdoc badge](https://cljdoc.org/badge/casa.squid/jack)](https://cljdoc.org/d/casa.squid/jack) [![Clojars Project](https://img.shields.io/clojars/v/casa.squid/jack.svg)](https://clojars.org/casa.squid/jack)
<!-- /badges -->

Jack Audio Connection Kit - Clojure API

## Features

`casa.squid.jack` makes it possible to connect to the Jack audio server, the de
facto standard for audio and music production on Linux, for audio, MIDI, and
transport control.

<!-- installation -->
## Installation

To use the latest release, add the following to your `deps.edn` ([Clojure CLI](https://clojure.org/guides/deps_and_cli))

```
casa.squid/jack {:mvn/version "0.2.12"}
```

or add the following to your `project.clj` ([Leiningen](https://leiningen.org/))

```
[casa.squid/jack "0.2.12"]
```
<!-- /installation -->

## Rationale

This library sprang forth from a desire for better MIDI handling on Linux, since
the built-in Java libaries `javax.sound.midi.*` were found lacking. Soon
transport handling and audio loop callback handling was added.

JNA Jack exposes the Jack API to Java/JVM, this libary makes it more
idiomatically consumable from Clojure. Particular care was taken to make it
REPL-friendly.

## Usage

Working with Jack requires that you first make a Jack client. `jack/client`
takes a string or keyword, and returns a client with the given name. This call
is memoized, so it's safe to call it multiple times, you'll get the same client
back.

```clj
(require '[casa.squid.jack :as jack])

(def c (jack/client :my-app))
```

Note that for many API calls passing in the client is optional, if omitted a
client with name `"Clojure"` is used.

Next you can create ports for this client, MIDI or audio, input or output.

```clj
(def midi-in (midi-in-port c :midi-in))
(def audio-left (audio-out-port c :left))
(def audio-right (audio-out-port c :right))
```

I have a little MIDI keyboard, let's find out what the port name is, and connect
it to our Jack client.

```clj
(jack/ports c #{:midi :out})
;; => ["Midi-Bridge:Midi Through:(capture_0) Midi Through Port-0"
;;     "Midi-Bridge:Launchkey Mini 2:(capture_0) Launchkey Mini LK Mini MIDI"
;;     "Midi-Bridge:Launchkey Mini 2:(capture_1) Launchkey Mini LK Mini InContro"]

(jack/connect "Midi-Bridge:Launchkey Mini 2:(capture_0) Launchkey Mini LK Mini MIDI" midi-in)
```

Now we can receive MIDI events. This has to happen inside a `:process` loop, so we register a process callback.

```clj
(jack/register
 c
 :process
 ::my-process-loop
 (fn [client frame-count]
   (run! (fn [[bytes frame]]
           (prn (midi/event bytes)))
         (jack/read-midi-events midi-in))
   true))
```

A few things to note here. `::my-process-loop` is a name we pick, if we call
this again with the same name, the previous callback will be replaced (compare
with `add-watch!`), making it convenient for REPL use. We can also `unregister`
the callback again.

The callback has to return `true`, returning falsey will cause subsequent
callbacks of the same type to be skipped.

`read-midi-events` returns a sequence of pairs, the MIDI event as a byte-array,
and a number indicating the frame offset of when this event occured within the
current processing cycle. The `casa.squid.midi` namespace has functions for
working with MIDI events, and converting them back and forth between byte arrays
and Clojure representation. (This namespace doesn't depend on anything Jack
specific, and may be split into its own mini-library in the future.)

Now if we press some keys or turn a knob, this shows up in the REPL:

```
;; channel / event type / note / velocity
[0 :note-on 59 72]
[0 :note-off 59 0]
[0 :note-on 59 69]
[0 :note-off 59 0]
;; channel / event type / controller number / value (0-127)
[0 :cc 23 47]
[0 :cc 23 48]
```

These are all the supported callbacks that you can register, with the arguments
of the callback function:

- :process [client frames]
- :buffer-size-changed [client buffersize]
- :client-registered [client name]
- :client-unregistered [client name]
- :ports-connected [client port-name-1 port-name-2]
- :ports-disconnected [client port-name-1 port-name-2]
- :port-registered [client port-name]
- :port-unregistered [client port-name]
- :sample-rate-changed [client rate]
- :client-shutdown [client]
- :update-position [client state frame pos new-pos]

### Dealing with Jack Transport

Jack Transport allows multiple applications to synchronize, reacting to the same
"play" or "pause" events, and agreeing on the current time signature, bar, beat,
etc. So you could for instance have a DAW containing audio recordings play in
sync to a step sequencer with drums.

Only one application at a time can be the transport leader. The leader is
responsible for computing the current musical position, based on the current
frame number. Note that not every leader will actually compute and set values
like bar, beat, tick, or beats-per-bar. They may be `0`, so be careful,
especially with divisions.

There are two ways to participate in Jack's transport and position. The simply
is to simply request the current position and transport state. You can do this
at any time, and from any thread (not just the processing thread).

```clj
(jack/transport-pos)
;; => {:frame 0, :frame-rate 44100, :usecs 17405102959, :valid #{}, :state :stopped}

;; Start rolling
(jack/start-transport!)

;; frame count is moving up, `:state` has changed to `:rolling`
(jack/transport-pos)
;;=>
{:frame 92160,
 :frame-rate 44100,
 :usecs 17493219351,
 :valid #{},
 :state :rolling}

;; Stop rolling
(jack/stop-transport!)

;; Seek back to the start (frame 0)
(jack/seek-transport! 0)
```

To become the transport leader you first call `make-transport-leader`, and then
register an `:update-position` callback.

`make-transport-leader` will fail if there is already a transport leader, unless
you pass a `force?` flag of true.

```clj
(jack/make-transport-leader true)
(jack/register
 client
 :update-position
 ::main-loop
 (fn [client state nframes ^JackPosition pos new-pos?]
   (populate-jack-pos pos (calculate-timings (.getFrameRate pos) (.getFrame pos) @timing))))
```

See `casa.squid.jack.transport-leader` for an example implementation.

```clj
(require '[casa.squid.jack.transport-leader :as l])

(l/initialize!)
```

You can see now that `jack/transport-pos` returns much richer timing
information, including bpm, bbt (bar/beat/tick), time signature
(beats-per-bar/beat-type).

```
(jack/transport-pos)
;;=>
{:valid #{:bbt},
 :tick 0,
 :beat-type 4.0,
 :beat 1,
 :frame-rate 44100,
 :beats-per-bar 4.0,
 :frame 0,
 :ticks-per-beat 1920.0,
 :bar 1,
 :state :stopped,
 :usecs 18019315664,
 :bpm 120.0,
 :bbt-offset 0}
```

### Producing audio

To generate audio, you write to the buffer of a given audio port during the
process callback. Here's an example that uses a precomputed wavetable of a sine
wave.

```clj
(def wt-size 100) ;; at an audio rate of 44100Hz this will produce a 441 Hz sine wave, i.e. a slightly sharp A.
(def wavetable (float-array wt-size))
(dotimes [i wt-size]
  (aset wavetable i (float (Math/sin (* 2 Math/PI (/ i wt-size))))))

;; Note that using `cycle` here will put some pressure on the garbage collector,
;; because a lot of Seq objects need to be created and discarded. It would be
;; preferable to keep track of indices in the wavetable manually. I'm doing it
;; this way to keep the implementation easy to understand.
(def wt-cycle (volatile! (cycle wavetable)))

(jack/register
 c
 :process
 ::sine-wave
 (fn [client frame-count]
   (let [buffer (.getFloatBuffer audio-left)]
     (.rewind buffer)
     (doseq [^float v (take frame-count @wt-cycle)]
       (.put buffer v))
     (vswap! wt-cycle #(drop frame-count %)))
   true))

;; What's cool is because the wavetable is a (mutable) float-array, you can
;; manipulate. To stop the sound, set it back to all zeroes.
(dotimes [i wt-size]
  (aset wavetable i (float 0)))
```

<!-- opencollective -->
<!-- /opencollective -->

<!-- contributing -->
<!-- /contributing -->

<!-- license -->
## License

Copyright &copy; 2023 Arne Brasseur and Contributors

Licensed under the term of the Mozilla Public License 2.0, see LICENSE.
<!-- /license -->


CHANGELOG.md
# Unreleased

## Added

## Fixed

## Changed

# 0.2.12 (2023-12-28 / 8bf5687)

## Added

## Fixed

## Changed

# 0.1.8 (2023-12-28 / 785a168)

- First release of `casa.squid.jack`

src/casa/squid/jack.clj
(ns casa.squid.jack
  "Wrapper for Jack (Jack Audio Connection Kit) midi

  Allows creating in/out ports, and registering callbacks of various types.

  Callback name + argument list of callback function

  - :process [client frames]
  - :buffer-size-changed [client buffersize]
  - :client-registered [client name]
  - :client-unregistered [client name]
  - :ports-connected [client port-name-1 port-name-2]
  - :ports-disconnected [client port-name-1 port-name-2]
  - :port-registered [client port-name]
  - :port-unregistered [client port-name]
  - :sample-rate-changed [client rate]
  - :client-shutdown [client]
  - :update-position [client state frame pos new-pos]

  See [[register]]/[[unregister]], which work analogous
  to [[add-watch!]] (idempotent, REPL safe, etc).
  "
  (:require
   [clojure.pprint :as pprint])
  (:import
   (java.util EnumSet)
   (org.jaudiolibs.jnajack
    Jack
    JackClient
    JackException
    JackMidi
    JackMidi$Event
    JackOptions
    JackPosition
    JackPositionBits
    JackPort
    JackPortFlags
    JackPortType
    JackProcessCallback
    JackShutdownCallback
    JackBufferSizeCallback
    JackPortConnectCallback
    JackPortRegistrationCallback
    JackSampleRateCallback
    JackShutdownCallback
    JackSyncCallback
    JackClientRegistrationCallback
    JackTimebaseCallback
    JackTransportState
    JackGraphOrderCallback
    JackStatus)))

(set! *warn-on-reflection* true)

;; The jack docs say to reuse a single instance... Might want to make this
;; thread-local or lock on it.
(defonce ^JackMidi$Event global-midi-event (JackMidi$Event.))
(defonce ^JackPosition global-jack-pos (JackPosition.))

(defonce !instance (delay (Jack/getInstance)))

(defn instance ^Jack [] @!instance)

(defonce
  ^{:doc "Registry of JackClientWrapper instances (keyword->JackClientWrapper).
  Used to make calls to [[client]] idempotent"}
  clients
  (atom {}))

(defprotocol Registry
  (register [this type key val] "Register a callback")
  (unregister [this type key] "Remove a callback")
  (lookup [this type key] "Find a given callback"))

(defprotocol PortId
  (port-name [this]))

(extend-protocol PortId
  String
  (port-name [this] this)
  JackPort
  (port-name [this] (.getName this)))

(defmulti init-callback!
  "When a callback of a given type is registered for the first time, register an
  actual Java callback object for that type, which dispatches to all of our
  registered callbacks."
  (fn [client type] type))

(defmethod init-callback! :default [_ _])

(def callback-pairs
  "Some pairs of callbacks types are bundled in a single Java callback interface,
  in these cases we need to know that if one of them is already initialized, the
  other is too."
  {:client-unregistered :client-registered
   :ports-disconnected :ports-connected
   :port-unregistered :port-registered})

(defrecord JackClientWrapper [^JackClient client registry]
  Registry
  (register [this t k value]
    (swap! registry
           (fn [registry]
             ;; FIXME: despite being inside the swap this can potentially still
             ;; cause issues when used at the same time from multiple threads,
             ;; since Clojure can retry this update function, and we want this
             ;; init to happen exactly once.
             (when-not (or (get registry t)
                           (get registry (get callback-pairs t))
                           (get registry (get (into {} (map (juxt val key)) callback-pairs) t)))
               (init-callback! this (get callback-pairs t t)))
             (assoc-in registry [t k] value))))
  (unregister [_ t k]
    (swap! registry update t dissoc k))
  (lookup [_ t k]
    (get-in @registry [t k]))

  ;; Convenience, deref to get the actual JackClient
  clojure.lang.IDeref
  (deref [_]
    @client)

  Object
  (toString [_]
    (str "#<JackClientwrapper " (.getName client) ">")))

(defmethod print-method JackClientWrapper [x ^java.io.Writer writer]
  (.append writer (.toString ^Object x)))

(defmethod pprint/simple-dispatch JackClientWrapper [x]
  (.write ^java.io.Writer *out* (.toString ^Object x)))

(defmacro registry-callback [client registry cb-type set-cb-type & methods]
  `(~(symbol (str "." set-cb-type))
    ~client
    (reify ~(symbol (str "Jack" cb-type "Callback"))
      ~@(for [[method k args] (partition 3 methods)]
          `(~method ~(into '[_] args)
            (reduce (fn [ok?# [k# cb#]]
                      ;; If a callback returns falsy, then any remaining
                      ;; callbacks of that type are skipped
                      (when ok?#
                        (try
                          (boolean (cb# ~@args))
                          (catch Exception e#
                            (println "Error in" '~method "callback" k#)
                            (println e#)
                            ok?#))))
                    true
                    (~k @~registry)))))
    ~@(drop (- (count methods) (mod (count methods) 3)) methods)))

(comment
  (macroexpand-1 '(registry-callback client registry ClientRegistration
                                     clientRegistered :client-registered [client name]
                                     clientUnregistered :client-unregistered [client name]))
  (require 'clojure.reflect)
  (clojure.reflect/reflect JackBufferSizeCallback))

(defn make-client
  "Construct a new Jack client. Prefer [[client]] which is idempotent."
  [name]
  (let [status (EnumSet/noneOf JackStatus)
        client (.openClient (instance)
                            name
                            (EnumSet/of JackOptions/JackNoStartServer)
                            status)
        registry (atom {})]
    #_(when (seq status)
        (println "make-client:" (map str status)))
    ;; These can't be registered once a client has been activated, so we do it
    ;; eagerly during init.
    (registry-callback client registry Process setProcessCallback
                       process :process [client frames])
    (registry-callback client registry BufferSize setBuffersizeCallback
                       buffersizeChanged :buffer-size-changed [client buffersize])
    (registry-callback client registry SampleRate setSampleRateCallback
                       sampleRateChanged :sample-rate-changed [client rate])
    (.activate client)
    (let [c (->JackClientWrapper client registry)]
      (swap! clients assoc name c)
      c)))

(defn client
  "Get a client for a given name, creating it if it doesn't exist."
  [client-name]
  (let [name (if (keyword? client-name)
               (subs (str client-name) 1)
               client-name)]
    (or (get @clients name) (make-client name))))

(defonce default-client (delay (client "Clojure")))

(defn jack-port-type ^JackPortType [kw]
  (case kw
    :audio JackPortType/AUDIO
    :midi JackPortType/MIDI
    ;; throws if kw is not recognized
    ))

(def port-flags
  {:can-monitor JackPortFlags/JackPortCanMonitor
   :in JackPortFlags/JackPortIsInput
   :out JackPortFlags/JackPortIsOutput
   :physical JackPortFlags/JackPortIsPhysical
   :terminal JackPortFlags/JackPortIsTerminal})

(defn jack-port-flags ^EnumSet [kws]
  (reduce (fn [^EnumSet es kw]
            (doto es (.add (get port-flags kw))))
          (EnumSet/noneOf JackPortFlags)
          kws))

(defn port
  "Get a virtual audio or midi port."
  ([name type flags]
   (port @default-client name type flags))
  ([^JackClientWrapper client name type flags]
   (assert (keyword? name))
   (if-let [port (lookup client type name)]
     port
     (let [port (.registerPort ^JackClient (:client client)
                               (subs (str name) 1)
                               (jack-port-type type)
                               (jack-port-flags flags))]
       (register client type name port)
       port))))

(defn midi-in-port
  "Get a virtual midi input port with a given name. Idempotent."
  ([name]
   (midi-in-port @default-client name))
  ([client name]
   (port client name :midi [:in])))

(defn midi-out-port
  "Get a midi output name for a given client with a given name. Idempotent."
  ([name]
   (midi-out-port @default-client name))
  ([client name]
   (port client name :midi [:out])))

(defn audio-in-port
  "Get a virtual audio input port with a given name. Idempotent."
  ([name]
   (audio-in-port @default-client name))
  ([client name]
   (port client name :audio [:in])))

(defn audio-out-port
  "Get a audio output name for a given client with a given name. Idempotent."
  ([name]
   (audio-out-port @default-client name))
  ([client name]
   (port client name :audio [:out])))

(defn read-midi-event
  "Read one of the midi events that are available in this processing cycle for the
  given port. Unless you're doing something very specific you probably want to
  use [[read-midi-events]] instead."
  [port idx]
  #_(locking global-midi-event)
  (JackMidi/eventGet global-midi-event port idx)
  (let [msg (byte-array (.size global-midi-event))]
    (.read global-midi-event msg)
    [msg (.time global-midi-event)]))

(defn read-midi-events
  "Read midi events that happened in this processing cycle for a given input port.
  Call within a processing callback."
  [port]
  (doall
   (for [idx (range (JackMidi/getEventCount port))]
     (read-midi-event port idx))))

(defn write-midi-event
  "Write a midi event to a given port at a given time (frame offset)."
  [port time msg]
  (JackMidi/eventWrite port time msg (count msg)))

(defn filter-pipe
  "Utility for creating midi filters, forward all messages from `in` to `out` if
  they satisfy `pred`."
  [in out pred]
  (try
    (JackMidi/clearBuffer out)
    (dotimes [idx (JackMidi/getEventCount in)]
      (let [[msg time] (read-midi-event in idx)]
        (when (pred msg)
          (write-midi-event out time msg))))
    true
    (catch JackException e
      (println "JackException:" e)
      true)))

(defn start-transport!
  "Start (play) Jack Transport control"
  ([]
   (start-transport! @default-client))
  ([client]
   (.transportStart ^JackClient (:client client))))

(defn stop-transport!
  "Stop Jack Transport control"
  ([]
   (stop-transport! @default-client))
  ([client]
   (.transportStop ^JackClient (:client client))))

(defn seek-transport!
  "Move the Jack Transport position to a specific frame."
  ([frame]
   (seek-transport! @default-client frame))
  ([client frame]
   (.transportLocate ^JackClient (:client client) frame)))

(defn ports
  "Get a vector of Jack ports (strings). Optionally takes a set of keywords to
  filter by type and port flags, e.g. #{:midi :out}, #{:audio :physical}.
  See [[port-flags]] for options."
  ([]
   (ports @default-client))
  ([client]
   (into [] (.getPorts (instance) (:client client) nil nil nil)))
  ([client type-and-flags]
   (into
    []
    (.getPorts
     (instance) (:client client) nil
     (cond
       (every? type-and-flags [:midi :audio])
       nil
       (:audio type-and-flags)
       JackPortType/AUDIO
       (:midi type-and-flags)
       JackPortType/MIDI
       :else nil)
     (let [flags (EnumSet/noneOf JackPortFlags)]
       (doseq [[kw flag] port-flags]
         (when (get type-and-flags kw)
           (.add flags flag)))
       flags)))))

(comment
  (ports (client :vibeflow) #{:midi :in}))

(defn connections
  "Get all jack connections as a sequence of `from`/`to` pairs (both String)."
  ([]
   (connections @default-client))
  ([client]
   (for [from (ports client #{:out})
         to (connections client from)]
     [from to]))
  ([client port]
   (into [] (.getAllConnections (instance) (:client client) port))))

(defn connect
  "Connect two jack ports, `from` and `two` are strings."
  ([conns]
   (doseq [[from to] conns]
     (connect from to)))
  ([from to]
   (connect @default-client from to))
  ([client from to]
   (.connect (instance) (:client client) (port-name from) (port-name to))))

(defn disconnect
  "Connect two jack ports, `from` and `two` are strings."
  ([from to]
   (disconnect @default-client from to))
  ([client from to]
   (.disconnect (instance) (:client client) (port-name from) (port-name to))))

(defn connect!
  "Sets jack connections to exactly the given connections, given as a list of
  pairs (port names, String). Will sever any existing connection that is not in
  the list. Ports that don't currently exist are ignored.

  Together with [[connections]] this gives you a way to capture the current
  state of a patch as EDN, and restore it later."
  ([conns]
   (connect! @default-client conns))
  ([client conns]
   (let [ports (ports client)
         conns (map (fn [[from to]] [(port-name from) (port-name to)])
                    conns)
         existing (set (for [from ports
                             to (connections client from)]
                         (vec (sort [from to]))))]
     (doseq [[from to] existing]
       (when-not (some #{[from to] [to from]} conns)
         (try
           (disconnect client to from)
           (catch Exception _))))
     (doseq [[from to] conns]
       (when-not (some #{[from to] [to from]} existing)
         (when (and (some #{from} ports) (some #{to} ports))
           (try
             (connect client from to)
             (catch Exception _))))))))

(def position-bits
  {JackPositionBits/JackPositionBBT :bbt
   JackPositionBits/JackPositionTimecode :timecode
   JackPositionBits/JackBBTFrameOffset :bbt-frame-offset
   JackPositionBits/JackAudioVideoRatio :audio-video-rate
   JackPositionBits/JackVideoFrameOffset :video-frame-offset})

(defn pos->map [^JackPosition pos]
  (let [valid (into #{} (map position-bits (.getValid pos)))]
    (cond-> {:frame (.getFrame pos)
             :frame-rate (.getFrameRate pos)
             :usecs (.getUsecs pos)
             :valid valid}
      (:bbt valid)
      (assoc
       :bar (.getBar pos)
       :beat (.getBeat pos)
       :tick (.getTick pos)
       :ticks-per-beat (.getTicksPerBeat pos)
       :bpm (.getBeatsPerMinute pos)
       :beats-per-bar (.getBeatsPerBar pos)
       :beat-type (.getBeatType pos)
       :bbt-offset 0)
      (:timecode valid)
      (assoc
       :frame-time (.getFrameTime pos)
       :bbt-offset (.getBbtOffset pos)
       :next-time (.getNextTime pos)
       :audio-frames-per-video-frame (.getAudioFramesPerVideoFrame pos)
       :video-offset (.getVideoOffset pos))
      (:bbt-frame-offset valid)
      (assoc
       :bbt-offset (.getBbtOffset pos)))))

(defn transport-pos
  "Get the current transport position as a map."
  ([]
   (transport-pos @default-client))
  ([client]
   (let [state (.transportQuery ^JackClient (:client client) global-jack-pos)]
     (assoc
      (pos->map global-jack-pos)
      :state
      (cond
        (= state JackTransportState/JackTransportStopped)     :stopped
        (= state JackTransportState/JackTransportRolling)     :rolling
        (= state JackTransportState/JackTransportLooping)     :looping
        (= state JackTransportState/JackTransportStarting)    :starting
        (= state JackTransportState/JackTransportNetStarting) :net-starting ;; Waiting for sync ready on the network
        )))))

(defmethod init-callback! :process [_ _])
(defmethod init-callback! :buffer-size-changed [_ _])
(defmethod init-callback! :sample-rate-changed [_ _])


(defmethod init-callback! :client-registered [{:keys [^JackClient client registry]} _]
  (registry-callback client registry ClientRegistration setClientRegistrationCallback
                     clientRegistered :client-registered [client name]
                     clientUnregistered :client-unregistered [client name]))

(defmethod init-callback! :ports-connected [{:keys [^JackClient client registry]} _]
  (registry-callback client registry PortConnect setPortConnectCallback
                     portsConnected :ports-connected [client port-name-1 port-name-2]
                     portsDisconnected :ports-disconnected [client port-name-1 port-name-2]))

(defmethod init-callback! :port-registered [{:keys [^JackClient client registry]} _]
  (registry-callback client registry PortRegistration setPortRegistrationCallback
                     portRegistered :port-registered [client port-name]
                     portUnregistered :port-unregistered [client port-name]))



(defmethod init-callback! :client-shutdown [{:keys [^JackClient client registry]} _]
  (registry-callback client registry Shutdown onShutdown
                     clientShutdown :client-shutdown [client]))

(defmethod init-callback! :graph-order-changed [{:keys [^JackClient client registry]} _]
  (registry-callback client registry GraphOrder setGraphOrderCallback
                     graphOrderChanged :graph-order-changed [client]))

;; Immediately throws, not clear why
;; (registry-callback client registry Sync setSyncCallback
;;                    syncPosition :sync-position [client position state])

(defn make-transport-leader
  "Attempt at becoming the Jack Transport leader, this requires you provide a
  `:update-position` callback, which takes the current transport state and
  frames, and based on that determines the other position fields, e.g
  beat/bar/tick.

  Will fail if another process is already leader, unless `force?` is set to
  true."
  ([]
   (make-transport-leader @default-client))
  ([client-or-force?]
   (make-transport-leader (if (boolean? client-or-force?)
                            @default-client
                            client-or-force?)
                          (if (boolean? client-or-force?)
                            client-or-force?
                            false)))
  ([client force?]
   (let [{:keys [ registry]} client]
     (registry-callback ^JackClient (:client client) registry Timebase setTimebaseCallback
                        updatePosition :update-position [client state frame pos new-pos]
                        (not force?)))
   client))

(defmethod init-callback! :update-position [client _]
  (make-transport-leader client false))


src/casa/squid/midi.clj
(ns casa.squid.midi
  "MIDI message parsing and construction logic.

  Interpreting and constructing of midi message as byte arrays (signed) or long
  arrays (unsigned).

  ``` clojure
  (def msg (byte-array 3 [130 64 100])) ;; Middle C on channel 2 with velocity 100

  (channel msg)    ;; => 2
  (event-type msg) ;; => :note-on
  (note msg)       ;; => 64
  (velocity msg)   ;; => 100

  (event msg) ;; => [2 :note-on 64 100]
  ```
  "
  (:require [clojure.set :as set])
  (:import (java.lang.reflect Array)))

(set! *unchecked-math* :warn-on-boxed)

;; https://www.songstuff.com/recording/article/midi_message_format/

;; https://en.wikipedia.org/wiki/MIDI_timecode
;; https://en.wikipedia.org/wiki/MIDI_beat_clock
;; https://en.wikipedia.org/wiki/MIDI_Machine_Control

(def event-types
  {0x80 :note-off
   0x90 :note-on
   0xA0 :poly-aftertouch
   0xB0 :cc
   0xC0 :program-change
   0xD0 :channel-aftertouch
   0xE0 :pitch-wheel
   0xF0 :sysex})

(def special-types
  {0xF0 :sysex-start
   0xF7 :sysex-end

   0xF2 :song-pointer
   0xF3 :song-select
   0xF6 :tune-request

   0xF1 :mtc-quarter-frame
   0xF8 :clock
   0xF9 :measure
   0xFA :start
   0xFB :continue
   0xFC :stop
   0xFE :active-sensing

   ;; Reset when sent over the wire, meta when read from a file
   0xFF :reset-meta

   0xF4 :unused
   0xF5 :unused
   0xFD :unused})

(defn b->l
  "Signed byte (-127 to 128) to long (0 to 255)"
  ^long [b]
  (long (bit-and (byte b) 0xFF)))

(defn l->b
  "Long to signed byte"
  [^long l]
  (unchecked-byte l))

(defmacro aget-byte [arr idx]
  `(byte (aget ~(with-meta arr {:tag 'bytes}) ~idx)))

;; (defn bytes->longs ^longs [^bytes bytes]
;;   (let [len (Array/getLength bytes)
;;         longs (long-array len)]
;;     (dotimes [idx len]
;;       (aset-long longs idx (b->l (aget bytes idx))))
;;     longs))

(defn status ^long [^bytes msg]
  (b->l (aget msg 0)))

(defn event-val ^long [msg]
  (bit-and 0xF0 (aget-byte msg 0)))

(defn special? [msg]
  (= 0xF0 (event-val msg)))

(defn event-type [msg]
  (if (special? msg)
    (get special-types (status msg))
    (get event-types (event-val msg))))

(defn channel ^long [msg]
  (bit-and 0x0F (aget-byte msg 0)))

(defn data1
  "Return the first data byte

  Semantics depend on the message type, e.g. note number, controller number,
  program number"
  ^long [msg]
  (bit-and 0xFF (aget-byte msg 1)))

(defn data2
  "Return the second data byte

  Semantics depend on the message type, e.g. velocity, pressure"
  ^long [msg]
  (bit-and 0xFF (aget-byte msg 2)))

(defn data-full
  "Return the data bytes combined into a single integer

  Used for pitch-wheel messages."
  ^long [msg]
  (bit-or (aget-byte msg 1) (bit-shift-left (aget-byte msg 2) 8)))

;; Semantic aliases, can be used in your code once you know the type of a
;; message to get more meaningful code.
(def note data1)
(def controller data1)
(def velocity data2)
(def cc-data data2)
(def pitchwheel data-full)

(defn event
  "Convert a message (array) to an event vector.

  Given a three element message array, return a four element vector with [channel
  event-type data-byte-1 data-byte-2]"
  [^bytes msg]
  (into [(channel msg) (event-type msg)] (rest msg)))

(def event-vals (into
                 (set/map-invert event-types)
                 (set/map-invert special-types)))

(defn message
  "Convert an event vector to a message array

  The inverse of [[event]]."
  ([v]
   (apply message v))
  ([channel type & data]
   (let [vals (cons (bit-or (long (event-vals type)) (long channel)) data)]
     (byte-array (count vals) vals))))

(comment
  (message 0 :note-on 64 127))


repl_sessions/walkthrough.clj
(ns walkthrough
  (:require
   [casa.squid.jack :as jack]
   [casa.squid.midi :as midi]))

(def c (jack/client :walkthrough))

(def midi-in (jack/midi-in-port c :midi-in))
(def audio-left (jack/audio-out-port c :left))
(def audio-right (jack/audio-out-port c :right))

(jack/ports c #{:midi :out})
;; => ["Midi-Bridge:Midi Through:(capture_0) Midi Through Port-0"
;;     "Midi-Bridge:Launchkey Mini 2:(capture_0) Launchkey Mini LK Mini MIDI"
;;     "Midi-Bridge:Launchkey Mini 2:(capture_1) Launchkey Mini LK Mini InContro"]

(jack/connect "Midi-Bridge:Launchkey Mini 2:(capture_0) Launchkey Mini LK Mini MIDI" midi-in)

(jack/register
 c
 :process
 ::my-process-loop
 (fn [client frame-count]
   (run! (fn [[bytes frame]]
           (prn (midi/event bytes)))
         (jack/read-midi-events midi-in))
   true))

(jack/transport-pos)
;; => {:frame 0, :frame-rate 44100, :usecs 17405102959, :valid #{}, :state :stopped}

(jack/start-transport!)

(jack/transport-pos)
;;=>
{:frame 92160,
 :frame-rate 44100,
 :usecs 17493219351,
 :valid #{},
 :state :rolling}

(jack/stop-transport!)
(jack/seek-transport! 0)


(require '[casa.squid.jack.transport-leader :as l])

(l/initialize!)

(jack/transport-pos)
;;=>
{:valid #{:bbt},
 :tick 0,
 :beat-type 4.0,
 :beat 1,
 :frame-rate 44100,
 :beats-per-bar 4.0,
 :frame 0,
 :ticks-per-beat 1920.0,
 :bar 1,
 :state :stopped,
 :usecs 18019315664,
 :bpm 120.0,
 :bbt-offset 0}


(jack/register
 c
 :process
 ::my-process-loop
 (fn [client frame-count]
   (run! (fn [[bytes frame]]
           (prn (midi/event bytes)))
         (jack/read-midi-events midi-in))
   true))

(def wt-size 100)
(def wavetable (float-array wt-size))
(dotimes [i wt-size]
  (aset wavetable i (float (Math/sin (* 2 Math/PI (/ i wt-size))))))

;; Note that using `cycle` here will put some pressure on the garbage collector,
;; because a lot of Seq objects need to be created and discarded. It would be
;; preferable to keep track of indices in the wavetable manually. I'm doing it
;; this way to keep the implementation easy to understand.
(def wt-cycle (volatile! (cycle wavetable)))

(jack/register
 c
 :process
 ::sine-wave
 (fn [client frame-count]
   (let [buffer (.getFloatBuffer audio-left)]
     (.rewind buffer)
     (doseq [^float v (take frame-count @wt-cycle)]
       (.put buffer v))
     (vswap! wt-cycle #(drop frame-count %)))
   true))

;; What's cool is because the wavetable is a (mutable) float-array, you can
;; manipulate. To stop the sound, set it back to all zeroes.
(dotimes [i wt-size]
  (aset wavetable i (float 0)))
